### ISSUE TITLE
Why fmt must use internet to fmt code？

### ISSUE THREAD

**Issue by driftluo** [rust-lang/rustfmt#3811]
rustfmt use `cargo-metadata` to get all kinds of metadata, but default use `online` mode and can not change to `offline`.

But it is well known that because of various network problems, `cargo` added the `--offline` option, `metadata` also supports this option, but `fmt` does not support.

I strongly recommend that `fmt` also have the offline parameter or default with `offline`.

https://github.com/rust-lang/rustfmt/blob/ee38d02591a2780a34e21313077a60b9a2a64ca0/src/cargo-fmt/main.rs#L511-L526

**Comment by calebcartwright** [comment]
Sounds reasonable to me. I'll work on adding this.

FWIW, `rustfmt` itself doesn't use `cargo_metadata`, so as a work around in the short term you could use `rustfmt` directly instead of `cargo fmt` if you're having network issues

**Comment by driftluo** [comment]
@calebcartwright @topecongiro   I think you don't understand what a bad network state is.

I have seen the final solution of pr, but from the results, it can only solve the situation that there is no network at all.

However, in China, a bad network does not mean that there is no network. In this case, the current #3813  cannot be solved.（Unless the network cable is disconnected or WiFi is turned off before calling `cargo fmt`, this is very unhumanized and stupid.）

The use of `cargo metadata` in China is as follows：

client request send ->  gfw -> server response -> gfw -- some data can't send to client -> client

Therefore, in the case of an appeal, the tcp access does not end, the client is waiting for a response, the server has sent a response message, but the message is intercepted by gfw, the client attempts to send, gfw continues to intercept。On Linux, default timeout seems 180 second，so client will wait util tcp shutdown, during this time, the user can't do anything.

Fallback offline #3813   is no practical help for the appeal. `Metadata` does not return an error immediately, but waits for the message to return, while `fmt` is waiting for a response from `metadata`. 

To simulate this situation is very simple, you can use the firewall to intercept the return message or start a server, sleep for more than 3 minutes for each request. This way, you may understand why I need the default offline or provide options.

---

Let's talk a little bit about the security of cargo not being connected to the Internet. 

I admit that under certain commands, if there is no default networking, there will be security issues, such as `cargo build`.

However, we must be clear that `fmt` is a lint tool that does not modify the source logic itself and does not output compilation results. Its function is limited to adjusting the code format.

The source file is local, it does not require any resources on the network. I think the reason why `fmt` calls `metadata` may be that it need to confirm the scope of the current project, such as workspace, and this work does not need to understand the source code of deps. I can't imagine that there will be security problems due to offline.

In summary, I think there is no problem with offline as the default choice.

**Comment by calebcartwright** [comment]
> @calebcartwright @topecongiro I think you don't understand what a bad network state is.

Hi @driftluo! Thanks for following up. First I just want to note that with any changes, it's really important to make sure that those changes don't break anything. The decision to stick with `online` mode as the default in #3813 was done out of caution for that very reason.

I can definitely understand how this issue is frustrating for you, and I would certainly like to help get this resolved! However, we do need to be careful that the solution for this particular use case doesn't cause _different_ problems for other users. 

IMO, using `offline` by default would not be ideal. The call to `cargo_metadata` has always used the default online mode, and AFAICT this is the first time an issue has been reported. Cargo/other cargo subcommands also use `online` by default, and it might be a little confusing for folks if `cargo fmt` did the exact opposite (regardless of the differences between the subcommands). In my experience, users typically expect to find similar behavior/options across cargo subcommands. 

> However, we must be clear that fmt is a lint tool that does not modify the source logic itself and does not output compilation results. 

Yup, we're aware 😉 

> I think the reason why fmt calls metadata may be that it need to confirm the scope of the current project, such as workspace, 

The `cargo fmt` subcommand does so to help simplify the process of running `rustfmt`, which is especially helpful in larger projects that have crates with multiple targets, multiple crates/workspace members, detecting editions, etc. As I mentioned in my original response, `rustfmt` itself does not

> and this work does not need to understand the source code of deps. I can't imagine that there will be security problems due to offline.

I noted this over in #3183, but I don't know the answer to this. I don't know enough about the `cargo` internals to know if for example, whether running `cargo fmt --all ...` in offline mode in a new environment (like a CI server) would be problematic, as the dep tree _is_ inspected when the `--all` flag is included.

I don't personally have any objections to switching back to adding the `--offline` flag that I originally submitted in #3813, though I do understand why @topecongiro wants be mindful about adding more and more CLI args. 

I'd be more than happy to make any additional changes to resolve this, but as always I defer to @topecongiro on which path to take.

**Comment by driftluo** [comment]
> running `cargo fmt --all` in  `offline` mode in a new environment 

I don't consider this, this may be have a problem. The possibilities should be carefully considered.

offline default fallback to inline？

**Comment by upsuper** [comment]
It's indeed a bit surprised that `cargo fmt` requires network by default, since it isn't supposed to need any build information at all. Even `cargo fmt --all` mentioned above doesn't seem to need network either, since its help message mentions "workspaces", which, I suppose, means formatting all packages under the current workspaces, which is still local.

@calebcartwright, do you think we should open a separate issue for further discussion of switching the default?

I can understand your concern that we may be breaking things if we don't take care. Are you aware of some way we can have people who have more knowledge around this involve in the discussion?

**Comment by calebcartwright** [comment]
> @calebcartwright, do you think we should open a separate issue for further discussion of switching the default?

I think it's fine to keep the discussion on this thread, at least for the time being

> It's indeed a bit surprised that cargo fmt requires network by default

The only reason `cargo fmt` has historically operated in `online` mode is because it needs to invoke `cargo metadata ...` commands (which it does via the `cargo_metadata` crate) in order to get the data it needs to pass as input to `rustfmt`, and `cargo` uses `online` mode by default

So the real connectivity question is about the `cargo metadata ...` commands. It is entirely possible (perhaps even probable) that `cargo metadata ...` will always work in every situation in `offline` mode, but I just personally don't know whether that is true 100% of the time. 

IMHO (for whatever it's worth 😄), going with an approach to solve this one particular use case that would change the default behavior for _every_ `cargo fmt` invocation for every user when there's still some unknowns/questions with that approach carries some degree of risk of unintended/undesireable side effects. 

> Even cargo fmt --all mentioned above doesn't seem to need network either, since its help message mentions "workspaces", which, I suppose, means formatting all packages under the current workspaces, which is still local

When the `--all` flag is included, there's a recursive stepping through the dep tree with multiple `cargo metadata` invocations. Again I'm not saying that `cargo metadata` needs network connectivity here, it was just one example of an instance where I don't know whether connectivity is needed, or what, if any, the impact of `offline` would be 🤷‍♂ 

https://github.com/rust-lang/rustfmt/blob/dbd89363913b2a35110077fed113aaa82b741079/src/cargo-fmt/main.rs#L379-L418

> Are you aware of some way we can have people who have more knowledge around this involve in the discussion

I imagine the folks on the cargo team would have the most insight into whether the `cargo metadata` command will always work in `offline` mode and/or whether there'd ever be any missing data elements in the `cargo metadata` results when using `offline` mode.

All that being said, how we proceed from here is up to @topecongiro, and I'm willing to help any way I can!

**Comment by calebcartwright** [comment]
I also went ahead and opened #3830 and #3831 with implementations of the alternative approaches to make it easier: one with `offline` by default with `online` fallback, and another that adds the `--offline` flag to `cargo fmt` to allow users to opt into `offline` mode.

**Comment by topecongiro** [comment]
@driftluo Thank you for elaborating on your concern.

I checked the source code of cargo. AFAICT,  `cargo metadata --no-deps` won't require any network connection whereas a plain `cargo metadata` may lead to a network connection. As in [the @calebcartwright's comment](https://github.com/rust-lang/rustfmt/issues/3811#issuecomment-538034326), the only usage of `cargo metadata` without the `--no-deps` option is when we run `cargo fmt --all`.

The usage of `cargo metadata` without the `--no-deps` option was introduced in https://github.com/rust-lang/rustfmt/pull/3664. Though I haven't confirmed, judging from [the comment](https://github.com/rust-lang/rustfmt/pull/3664/files#discussion_r298856004), we can remove this usage by updating `cargo_metadata`. I prefer to take this route, if possible, as it does not require adding a command line option while keeping cargo-fmt completely offline.

@calebcartwright Thank you for creating #3830 and #3831. As I noted above, I prefer to try removing the usage of `cargo metadata` without the `--no-deps` first, but if that turns out to be infeasible, then I am happy to merge one of two of your PRs.

**Comment by calebcartwright** [comment]
> Though I haven't confirmed, judging from the comment, we can remove this usage by updating cargo_metadata

That would require a patch to `cargo` itself wouldn't it? AFACIT the `cargo_metadata` crate is primarily just running `cargo metadata ..` commands and deserializing the command output. 

We'd need to cargo to be updated so that running `cargo metadata --no-deps` would include the `manifest_path` value for each dependency, and additionally, for any local dependencies, cargo would also have to recursively traverse their dependencies to do the same (to support nested/hierarchical local dependencies) to prevent a recurrence of the bug in #3643
